sdk/dotnet/Pulumi.Tests/Serialization/EnumConverterTests.cs

```diff
@@ -0,0 +1,214 @@
+// Copyright 2016-2020, Pulumi Corporation
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Threading.Tasks;
+using Pulumi.Serialization;
+using Xunit;
+
+namespace Pulumi.Tests.Serialization
+{
+    public class EnumConverterTests : ConverterTests
+    {
+        [EnumType]
+        public readonly struct ContainerColor : IEquatable<ContainerColor>
+        {
+            private readonly string _value;
+
+            private ContainerColor(string value)
+            {
+                _value = value ?? throw new ArgumentNullException(nameof(value));
+            }
+
+            public static ContainerColor Red { get; } = new ContainerColor("red");
+            public static ContainerColor Blue { get; } = new ContainerColor("blue");
+            public static ContainerColor Yellow { get; } = new ContainerColor("yellow");
+
+            public static bool operator ==(ContainerColor left, ContainerColor right) => left.Equals(right);
+            public static bool operator !=(ContainerColor left, ContainerColor right) => !left.Equals(right);
+
+            public static explicit operator string(ContainerColor value) => value._value;
+
+            [EditorBrowsable(EditorBrowsableState.Never)]
+            public override bool Equals(object? obj) => obj is ContainerColor other && Equals(other);
+            public bool Equals(ContainerColor other) => string.Equals(_value, other._value, StringComparison.Ordinal);
+
+            [EditorBrowsable(EditorBrowsableState.Never)]
+            public override int GetHashCode() => _value?.GetHashCode() ?? 0;
+
+            public override string ToString() => _value;
+        }
+
+        [EnumType]
+        public readonly struct ContainerBrightness : IEquatable<ContainerBrightness>
+        {
+            private readonly double _value;
+
+            private ContainerBrightness(double value)
```

Comment:
I assume we want to support float-based enums (given the discussions we've had about the other languages).

Note that if that's the case, we'll need to specifically support `double` in the codegen. The emitted code is subtly different than for `string` (e.g. the constructor doesn't throw, `Equals` doesn't use `string.Equals`, `GetHashCode` doesn't have any null checks, and `ToString` calls `ToString()` on the `_value`).

Reply:
Have we discussed/seen a need for enums based on a boolean?

Reply:
Boolean-based enums are supported by the schema (and nodejs) but I haven't seen them in practice anywhere yet. Are you aware of such enums in azure @mikhailshilkov? On the codegen side, C# will throw an unimplemented error if bool-based enums are present in the schema. I can create an issue and add it as a comment in the codegen unless you feel strongly that it should be implemented immediately.

Reply:
No, I haven't seen them yet. An issue sounds good.